Pro revision

2bndy5 opened the pull request:
## Changes include
- reduced the duplicate code
- abstracted some functionality into separate functions
- improved and added unit tests (also added a .coveragerc file to configure coverage measurement)
- added creation of a symlink if possible.
    - symlink will not overwrite anything that already exists
    - added a `-f` option so that users can force overwriting an existing symlink
    - symlink will not be created if the symlink name is already occupied by an existing file that is not a symlink (`-f` will also fail in this case)
- changed call to `subprocess.run(["tar", "-xvf"])` to use `tarfile` module (part of python std lib).
   - technically, this function is dead code, but it might be useful in the future.
- added `urllib.error.HTTPError` to list of caught exceptions when downloading a file.
- added some verbosity with `print()` statements. This can still be improved, but I just wanted to see some conditional output.
- updated the install config to use a pyproject.toml file. Using setup.py alone is no longer the convention, and will likely be deprecated in the future (setuptools issues a deprecation warning about this already).
- add a step in the publishing CI workflow to upload to test.PyPI when the workflow is triggered by a `workflow_dispatch`
   - The CI will now only publish binary wheels (`.whl` files) instead of both binary wheel and software distribution (the tar.gz archive). This won't have an impact on installs as they were already using the binary wheels (fastest method) which are compatible with any machine because this is written in pure python (no c-extensions).
   - switched to using `twine` directly for both checking and uploading the built wheel.
- The test CI workflow now builds a single binary wheel which is used in the install job. It runs a lot faster now. And this also behaves the same as it would in the cpp-linter-action CI (when this is merged)

shenxianpeng commented on clang_tools/__init__.py:
```diff
@@ -1,4 +1,7 @@
+RESET_COLOR = "\033[0m"
+YELLOW = "\033[93m"
 install_os = check_install_os()
+suffix = ".exe" if install_os == "windows" else ""
```
```suggestion
suffix = ".exe" if install_os == "windows" else ""

```

shenxianpeng commented:
![image](https://user-images.githubusercontent.com/3353385/183272345-21114b8a-c4ed-44eb-93a5-1a301967567b.png)
The added bin file `tests/file.tar.gz` looks a little big, is this file used for testing?

shenxianpeng reviewed (approved):
All the revisions are awesome!! open source code is probably better than enterprise code :)

shenxianpeng commented:
The current revsion is not formatted by pre-commit, I have introduced [.pre-commit-config.yaml](https://github.com/cpp-linter/clang-tools-pip/blob/main/.pre-commit-config.yaml), enabled pre-commit can help format code automatically when running `git commit command. 

To format all the code can run the fellow commands:


```bash
ubuntu@CN-L-2680:~/clang-tools-pip$ pre-commit install
pre-commit installed at .git/hooks/pre-commit
ubuntu@CN-L-2680:~/clang-tools-pip$ pre-commit run --all-files
Trim Trailing Whitespace.................................................Passed
Fix End of Files.........................................................Passed
Check docstring is first.................................................Passed
Check for added large files..............................................Passed
Check Yaml...............................................................Passed
Debug Statements (Python)................................................Passed
Fix requirements.txt.....................................................Passed
Reorder python imports...................................................Failed
- hook id: reorder-python-imports
- exit code: 1
- files were modified by this hook

Reordering imports in tests/test_install.py
Reordering imports in clang_tools/util.py
Reordering imports in tests/test_main.py
Reordering imports in clang_tools/main.py
Reordering imports in clang_tools/install.py
Reordering imports in tests/test_util.py

pyupgrade................................................................Passed
flake8...................................................................Failed
- hook id: flake8
- exit code: 1

clang_tools/install.py:16:1: E302 expected 2 blank lines, found 1

tests....................................................................
```

2bndy5 commented:
> The added bin file `tests/file.tar.gz` looks a little big, is this file used for testing?

yes, it is used to test the `unpack()` function.

---
I'll enable pre-commit locally and push the necessary changes.

2bndy5 commented:
The tests (& and test assets) should not be distributed to pypi. That wouldn't make sense if they are.

2bndy5 commented:
I don't think the tests should be run with pre-commit because it is not a quick task.
```
============= 46 passed, 3 xfailed in 16.57s ===========
```
It just seems unusual to me because I've never seen tests performed in the pre-commit stage. The local hook "tests" fails for me because I'm using windows. now I can't push anything.

shenxianpeng commented:
Looks like the tests takes more time than before. Here is the output of main branch.

```bash
ubuntu@CN-L-2680:~/clang-tools-pip$ pre-commit run --all-files
Trim Trailing Whitespace.................................................Passed
Fix End of Files.........................................................Passed
Check docstring is first.................................................Passed
Check for added large files..............................................Passed
Check Yaml...............................................................Passed
Debug Statements (Python)................................................Passed
Fix requirements.txt.....................................................Passed
Reorder python imports...................................................Passed
pyupgrade................................................................Passed
flake8...................................................................Passed
tests....................................................................Passed
ubuntu@CN-L-2680:~/clang-tools-pip$ git branch
* main
```

2bndy5 commented:
Yep, that's what happens with more coverage and more features... They also use temporary directories which seem to be seen by pytest as malformed paths when executed in pre-commit (at least on windows).

2bndy5 commented:
I updated the build/install/publish workflows and created docs with sphinx-immaterial theme (the project that had me distracted for so long).

The setup.py file is a dummy in favor of using pyproject.toml (the more modern standard convention). I also added some mypy config to the pyproject.toml (I think flake8 also supports using this file as a config).

2bndy5 commented:
I published a test [release to test.pypi](https://test.pypi.org/project/clang-tools/) just to be sure I didn't break anything.

2bndy5 commented:
I've been testing these revisions (using the release from testPyPI) on the cpp-linter-action/backlogged-updates branch with success. 👍🏼

shenxianpeng commented on README.rst:
```diff
@@ -0,0 +1,110 @@
+
+.. code-block:: shell
+
+    clang-tools --help
```
The new usage should be 

```bash
usage: clang-tools [-h] [-i INSTALL] [-d DIRECTORY] [-f]

optional arguments:
  -h, --help            show this help message and exit
  -i INSTALL, --install INSTALL
                        Install clang-tools with specific version. default is 13.
  -d DIRECTORY, --directory DIRECTORY
                        The directory where is the clang-tools install.
  -f                    Force overwriting the symlink to the installed binary. This will only overwrite an existing
                        symlink.
```

2bndy5 replied:
good catch! I updated it.

shenxianpeng reviewed (changes requested):
The "Supported versions" section on the doc page doesn't show the table well:
https://cpp-linter.github.io/clang-tools-pip/

How can find the development document generated by docstring on github-pages?

2bndy5 commented on .github/workflows/python-publish.yml:
```diff
@@ -23,18 +23,29 @@ jobs:
     - uses: actions/checkout@v3
+      # use fetch --all for setuptools_scm to work
+      with:
+        fetch-depth: 0
```
setuptools_scm needs the full history to create an adequate version. My first release to test-pypi was `v0.2.4.post1.dev8` (did this manually from my local machine. The second release to test-pypi (using the CI workflow_dispatch event) ended up publishing `v0.0.post1.dev1`, but it should've been `v0.2.4.post1.dev13`. After adding this `fetch-depth: 0`, the third release to test-pypi was accurately named `v0.2.4.post1.dev14`.

You can see the [versions published to test-pypi here](https://test.pypi.org/project/clang-tools/#history), but the timestamp isn't very descriptive.

2bndy5 replied:
you can also see the wheel built by the test CI by [downloading the wheel artifact](https://github.com/cpp-linter/clang-tools-pip/actions/runs/2813802729#artifacts). Upon unzipping the download, you'll see that the version is actually inaccurate, but it doesn't matter for the test CI.

2bndy5 commented:
> The "Supported versions" section on the doc page doesn't show the table well: https://cpp-linter.github.io/clang-tools-pip/
> 
> How can find the development document generated by docstring on github-pages?

That page is based on the main branch and will be updated when this is merged. To see what the docs will look like, you can download [the docs artifact from the test CI](https://github.com/cpp-linter/clang-tools-pip/actions/runs/2813802729#artifacts).

That page was initially generated because I created the gh-pages branch and github just used the main branch without using these changes. The problem with that table is probably from not putting a blank line between the section title and the table:
https://github.com/cpp-linter/clang-tools-pip/blob/a63ca48822d0e3092686f26ec40162f75e936b9b/README.md?plain=1#L53-L54
Again the error is from the main branch, not this branch.

shenxianpeng commented:
> The tests (& and test assets) should not be distributed to pypi. That wouldn't make sense if they are.

Even we don't distribute tests to pypi, we should not upload large assets to the git repo. we have a hook to check added large file.

https://github.com/cpp-linter/clang-tools-pip/blob/a63ca48822d0e3092686f26ec40162f75e936b9b/.pre-commit-config.yaml#L10

> > The added bin file `tests/file.tar.gz` looks a little big, is this file used for testing?
> 
> yes, it is used to test the `unpack()` function.

unpack_file() function seems not to be used at this moment, if you could not make it to a fake file or very smaller, maybe could remove both.

2bndy5 commented:
I'm ok with removing the dead codfe (& related test). It wouldn't be hard to add this functionality in the future... I have some extensive experience working with compressed archives using python std libs.

shenxianpeng reviewed (approved):
LGTM

shenxianpeng commented:
Publish the package to PyPI step is skipped by the fellowing contion


```
if: github.event_name == 'release' && github.repository == 'cpp-linter/clang-tools-pip' && github.ref == 'refs/heads/main'
```

Maybe the event_name is published not release?

https://github.com/cpp-linter/clang-tools-pip/runs/7720622852?check_suite_focus=true

2bndy5 commented:
>Maybe the event_name is published not release?

Yeah, that might be it. ([github docs](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#release)).

shenxianpeng commented:
From the log

```
2022-08-08T07:20:51.2100210Z Job defined at: cpp-linter/clang-tools-pip/.github/workflows/python-publish.yml@refs/tags/v0.3.0
```

Maybe because `&& github.ref == 'refs/heads/main'`?

shenxianpeng commented:
This seems the event name

![image](https://user-images.githubusercontent.com/3353385/183364781-c4deafa0-6204-41c9-858b-e0278b47c224.png)

shenxianpeng commented:
I want to remove `&& github.ref == 'refs/heads/main'` and have another try if you have no objection.

2bndy5 commented:
> Maybe because && github.ref == 'refs/heads/main'?

I've used that on other repos. It just makes sure that releases are only uploaded from the main branch.

I think we're fighting the difference between enevt_name and event_type ([github actions context](https://docs.github.com/en/actions/learn-github-actions/contexts#github-context))

2bndy5 commented:
I changed it to `event_name == 'published`. you'll have to delete the release and re-publish it using the same tag.

shenxianpeng commented:
OK

2bndy5 commented:
🤞🏼 sorry.

2bndy5 commented:
Damn it!

2bndy5 commented:
hold on I'm going to try something.

There was no way to test this part without going through this...

shenxianpeng commented:
Sure. maybe the branch name is not called main instead of tag in this release workflow

2bndy5 commented:
We use the same condition on the sphinx-immaterial project, so I don't think that condition is the problem.
https://github.com/jbms/sphinx-immaterial/blob/bf169751fba59362bcd06da96bfa65b2728cbbae/.github/workflows/build.yml#L122

2bndy5 commented:
oh, I forgot to fast-forward the tag, so none of my attempts were having any affect.

shenxianpeng commented:
They are not the same conditon.
`if: ${{ github.event_name == 'push' && github.repository == 'jbms/sphinx-immaterial' && (startsWith(github.ref, 'refs/tags/v') || github.ref == 'refs/heads/main') }}`

`(startsWith(github.ref, 'refs/tags/v') || github.ref == 'refs/heads/main')` it will still ture if the branch is a tag not main

2bndy5 commented:
I'm removing that condition and adding 
```yml
on:
  release:
    branches: [main]
    types: [published]
  workflow_dispatch:

# ...
    - name: Publish package (to PyPI)
      if: github.event_name != 'workflow_dispatch' && github.repository == 'cpp-linter/clang-tools-pip'
```

shenxianpeng commented:
IAnd I guess this still work.`if: github.event_name = 'release' && github.repository == 'cpp-linter/clang-tools-pip'`

2bndy5 commented:
That worked! https://pypi.org/project/clang-tools/0.3.0/

2bndy5 commented:
It would seem there are too many ways to do it. Feel free to change it. I want to make sure that `workflow_dispatch` events do not publish to pypi. rather, `workflow_dispatch` events should only publish to test-pypi.

2bndy5 commented:
@shenxianpeng If you have a test-pypi account (which is different from actual pypi account), then I'll invite you to the test releases' projects on test-pypi. I'm getting [cpp-linter-action ready to publish to pypi as well](https://test.pypi.org/project/cpp-linter/)

2bndy5 commented:
![image](https://user-images.githubusercontent.com/14963867/183375740-0f737975-5451-47b4-b343-4b38ab0f1c11.png)
So, I copied and pasted much from this project's pyproject.toml 🤣 I've already corrected the image and brief description inaccuracies on the backlogged-updates branch, but those changes haven't been released to the testpypi project yet.

shenxianpeng commented:
My testpypi account is also xpshen
